pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitSlicedRangeIndexReader.java

```diff
@@ -0,0 +1,171 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.segment.index.readers;
+
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import javax.annotation.Nullable;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.index.reader.RangeIndexReader;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.roaringbitmap.RangeBitmap;
+import org.roaringbitmap.RoaringBitmap;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+
+import static org.apache.pinot.spi.data.FieldSpec.DataType.DOUBLE;
+import static org.apache.pinot.spi.data.FieldSpec.DataType.FLOAT;
+
+public class BitSlicedRangeIndexReader implements RangeIndexReader<ImmutableRoaringBitmap> {
+
+  @SuppressWarnings("rawtypes")
+  private static final AtomicReferenceFieldUpdater<BitSlicedRangeIndexReader, SoftReference> UPDATER =
+      AtomicReferenceFieldUpdater.newUpdater(BitSlicedRangeIndexReader.class, SoftReference.class,
+          "_rangeBitmapRef");
+
+  private final PinotDataBuffer _dataBuffer;
+  private final long _offset;
+  private final long _min;
+  private final double _scale;
+  private volatile SoftReference<RangeBitmap> _rangeBitmapRef;
+
+  public BitSlicedRangeIndexReader(PinotDataBuffer dataBuffer, ColumnMetadata metadata) {
```

Comment:
Dictionary encoded column is not handled in the reader

Reply:
It is supported by querying for `int` ranges. For dictionary encoded columns, the `RangeBitmap` will have been built over dictionary ids, as `BitSlicedRangeIndexCreator` implements `DictionaryBasedInvertedIndexCreator`. 

Are you concerned that double/float columns might have a dictionary (I would not expect this configuration)?

Reply:
Double/float columns can have dictionary. Sometimes these columns can have low cardinality, and dictionary-encoding gives better performance.

Reply:
I think this is now correct, the reader doesn't care what the data type is and no longer depends on `ColumnMetadata`. It expects to be queried with the right types (ints for dictionarized or raw int columns, long, float, or double for LONG, FLOAT, or DOUBLE raw columns. 

If the caller can't/shouldn't be trusted to provide column-type appropriate arguments, I can put type checking/conversions in.